### PR TITLE
feat(hybrid-cloud): Add customer domain support to slack unfurls

### PR DESCRIPTION
**Pull request by dashed**
Add customer domain support to Slack unfurls. For example, https://org1.sentry.io/issues/12345/ should unfurl in Slack.

### CODE REVIEW

File: src/sentry/integrations/slack/unfurl/__init__.py [hunk 1]
```diff
@@ -44,7 +44,9 @@ def type_coercer(url: str, args: Mapping[str, str | None]) -> Mapping[str, Any]:
 from .issues import handler as issues_handler
 from .metric_alerts import handler as metric_alert_handler
 
+# This
 link_handlers = {
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/__init__.py:48]: What are the comments for?

File: src/sentry/integrations/slack/unfurl/__init__.py [hunk 2]
```diff
@@ -53,7 +55,15 @@ def type_coercer(url: str, args: Mapping[str, str | None]) -> Mapping[str, Any]:
 
 def match_link(link: str) -> tuple[LinkType | None, Mapping[str, Any] | None]:
     for link_type, handler in link_handlers.items():
-        match = handler.matcher.match(link)
+        # THis
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/__init__.py:58]: Same here.

File: src/sentry/integrations/slack/unfurl/discover.py [hunk 3]
```diff
@@ -298,10 +298,16 @@ def map_discover_query_args(url: str, args: Mapping[str, str | None]) -> Mapping
     return dict(**args, query=query)
 
 
+discover_link_regex = re.compile(
+    r"^https?\://(?#url_prefix)[^/]+/organizations/(?P<org_slug>[^/]+)/discover/(results|homepage)"
+)
+
+customer_domain_discover_link_regex = re.compile(
+    r"^https?\://(?P<org_slug>[^/]+?)\.(?#url_prefix)[^/]+/discover/(results|homepage)"
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/discover.py:306]:
```suggestion
    r"^https?\://(?P<org_slug>[^.]+?)\.(?#url_prefix)[^/]+/discover/(results|homepage)"
```

You'll hit a `.` before you'll hit a `/` in the domain name.

File: src/sentry/integrations/slack/unfurl/issues.py [hunk 4]
```diff
@@ -63,10 +65,16 @@ def unfurl_issues(
     return out
 
 
+issue_link_regex = re.compile(
+    r"^https?\://(?#url_prefix)[^/]+/(?#organizations)[^/]+/(?#organization_slug)[^/]+/issues/(?P<issue_id>\d+)(?:/events/(?P<event_id>\w+))?"
+)
+
+customer_domain_issue_link_regex = re.compile(
+    r"^https?\://(?#url_prefix)[^/]+/issues/(?P<issue_id>\d+)(?:/events/(?P<event_id>\w+))?"
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/issues.py:73]:
```suggestion
    r"^https?\://(?#url_prefix)[^.]+/issues/(?P<issue_id>\d+)(?:/events/(?P<event_id>\w+))?"
```

You'll hit a `.` before you hit a `/` here too.

File: src/sentry/integrations/slack/unfurl/issues.py [hunk 5]
```diff
@@ -63,10 +65,16 @@ def unfurl_issues(
     return out
 
 
+issue_link_regex = re.compile(
+    r"^https?\://(?#url_prefix)[^/]+/(?#organizations)[^/]+/(?#organization_slug)[^/]+/issues/(?P<issue_id>\d+)(?:/events/(?P<event_id>\w+))?"
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/issues.py:69]:
```suggestion
    r"^https?\://(?#url_prefix)[^/]+/organizations/(?#organization_slug)[^/]+/issues/(?P<issue_id>\d+)(?:/events/(?P<event_id>\w+))?"
```

If you want to match `organizations` specifically you could use a literal here.

File: src/sentry/integrations/slack/unfurl/metric_alerts.py [hunk 6]
```diff
@@ -137,10 +137,16 @@ def map_metric_alert_query_args(url: str, args: Mapping[str, str | None]) -> Map
     return map_incident_args(url, data)
 
 
+metric_alerts_link_regex = re.compile(
+    r"^https?\://(?#url_prefix)[^/]+/organizations/(?P<org_slug>[^/]+)/alerts/rules/details/(?P<alert_rule_id>\d+)"
+)
+
+customer_domain_metric_alerts_link_regex = re.compile(
+    r"^https?\://(?P<org_slug>[^/]+?)\.(?#url_prefix)[^/]+/alerts/rules/details/(?P<alert_rule_id>\d+)"
```
Comment (markstory) [inline src/sentry/integrations/slack/unfurl/metric_alerts.py:145]:
```suggestion
    r"^https?\://(?P<org_slug>[^.]+?)\.(?#url_prefix)[^/]+/alerts/rules/details/(?P<alert_rule_id>\d+)"
```

**Review by markstory — APPROVED** [review]

File: src/sentry/integrations/slack/unfurl/issues.py (hunk 4 above)
Comment (dashed) [reply src/sentry/integrations/slack/unfurl/issues.py:73]: This suggestion is not applicable here since we're not matching on an org slug subdomain.

**Pull request merged**